azurelinuxagent/common/protocol/extensions_config_retriever.py

```diff
@@ -0,0 +1,374 @@
+GOAL_STATE_SOURCE_FILE_NAME = "GoalStateSource"
+VM_ID_FILE_NAME = "VmId"
+
+GOAL_STATE_SOURCE_FABRIC = "Fabric"
+GOAL_STATE_SOURCE_FASTTRACK = "FastTrack"
+
+DMIDECODE_CALL = "dmidecode"
+
+
+class ExtensionsConfigReasons:
+    FABRIC_CHANGED = "FabricChanged"
+    FABRIC_LAST_CHANGE = "LastFabric"
+    FAST_TRACK_CHANGED = "FTChanged"
+    FAST_TRACK_LAST_CHANGE = "LastFT"
+
+
+class FastTrackChangeDetail:
+    NO_CHANGE = "NoChange"
+    NO_EXTENSIONS = "NoExtensions"
+    NO_PROFILE = "NoProfile"
+    SEQ_NO_CHANGED = "seqNoChanged"
+
+
+class FabricChangeDetail:
+    INCARNATION_CHANGED = "IncChanged"
+    NO_CHANGE = "NoChange"
+    NO_INCARNATION = "NoInc"
+
+
+""" 
+GenericExtensionsConfig abstracts whether we pulled the goal state from Fabric or from FastTrack
+consumers should not worry from where the ExtensionsConfig came. They should also have no knowledge
+of sequence numbers or incarnations, which are specific to FastTrack and Fabric respectfully
+"""
+class GenericExtensionsConfig(ExtensionsConfig):
+    def __init__(self, extensions_config, changed, ext_conf_retriever):
+        self.extensions_config = extensions_config
+        self.changed = changed
+        self._ext_conf_retriever = ext_conf_retriever
+
+        ExtensionsConfig.__init__(self, extensions_config.xml_text)
+
+    def commit_processed(self):
+        self._ext_conf_retriever.commit_processed()
+
+    def get_description(self):
+        return self._ext_conf_retriever.get_description()
+
+class ExtensionsConfigRetriever(object):
+    def __init__(self, wire_client):
+        self._wire_client = wire_client
+        self._last_fabric_incarnation = None
+        self._last_svd_seq_no = None
+        self._last_fast_track_seq_no = None
+        self._saved_artifacts_profile = None
+        self._last_mode = None
+        self._pending_mode = None
+        self._pending_fast_track_seq_no = None
+        self._pending_svd_seq_no = None
+        self._pending_fabric_incarnation = None
+        self._fast_track_changed_detail = None
+        self._fabric_changed_detail = None
+        self._reason = None
+        self._reset_if_necessary()
+
+    def get_ext_config(self, incarnation, ext_conf_uri):
+        # If we don't have a uri, return an empty extensions config
+        is_startup = False
+        if ext_conf_uri is None:
+            return GenericExtensionsConfig(ExtensionsConfig(None), False, self)
+
+        # Get the Fabric goal state and whether it changed
+        fabric_changed = self._get_fabric_changed(incarnation)
+
+        # Get the VmArtifactsProfile and whether fast track changed, if enabled
+        artifacts_profile, fast_track_changed = self._get_fast_track_details()
+
+        self._pending_mode = self._decide_what_to_process(fabric_changed, fast_track_changed)
+        if self._last_mode is None:
+            is_startup = True
+            logger.info("Processing first mode {0}. Reason={1}", self._pending_mode, self._reason)
```

Comment:
Could we re-word this message a little?
> Processing first mode FastTrack. Reason=LastFT: FastTrack=NoChange, Fabric=NoChange

Assuming I am reading it correctly, may be something similar to

> Using FastTrack for the first call to extensions. Reason=LastFT: FastTrack=NoChange, Fabric=NoChange

And maybe use "LastFastTrack" instead of "LastFT"?

Reply:
Changed to LastFastTrack and re-worded the message.